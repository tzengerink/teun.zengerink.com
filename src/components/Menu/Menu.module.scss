@import '../../styles/index';

.menu {
  h1 {
    position: fixed;
    z-index: 4;
    top: $margin;
    left: $margin;
    padding-top: 4px;
  }
}

.navigation {
  @include transition(0.1s);
  @include transform(translateY(-50px));

  position: fixed;
  top: 0;
  left: $margin-small;
  bottom: 0;
  visibility: hidden;
  opacity: 0;
  z-index: 3;
  padding: 60px $margin $margin $margin;
  width: 100%;
  background: $background-color-transparent;

  h2 {
    margin-top: 30px;
    margin-bottom: 10px;
    cursor: default;
  }

  ul {
    list-style: none;
    padding: 0;
  }

  li {
    a {
      font-size: 0.8em;
    }

    &.active {
      text-decoration: underline;
    }
  }
}

.hamburger {
  display: block;
  z-index: 4;
  position: fixed;
  top: $margin;
  right: $margin;
  cursor: pointer;
}

.line {
  @include transition(0.3s);

  display: block;
  width: 24px;
  height: 3px;
  margin: 4px auto;
  background-color: $text-color;
}

.socials {
  position: absolute;
  right: $margin;
  padding-right: $margin;
  margin-right: $margin;
  bottom: $margin;

  a {
    text-decoration: none;
  }

  img {
    float: left;
    width: 22px;
    height: 22px;
    opacity: 0.7;

    &:hover {
      opacity: 1;
    }
  }
}

.open {
  .navigation {
    @include transform(none);

    visibility: visible;
    opacity: 1;
  }

  .hamburger .line:nth-child(1) {
    @include transform(translateY(8px) rotate(45deg));
  }

  .hamburger .line:nth-child(2) {
    opacity: 0;
  }

  .hamburger .line:nth-child(3) {
    @include transform(translateY(-6px) rotate(-45deg));
  }
}

@include from-tablet-portrait {
  .menu {
    h1 {
      padding-top: 0;
    }
  }

  .navigation {
    @include transform(none);

    position: fixed;
    opacity: 1;
    visibility: visible;
    right: $margin;
    bottom: auto;
    left: auto;
    width: auto;
    padding: 0;
    text-align: right;

    h2 {
      margin-top: $margin;
      margin-bottom: 0;
      text-transform: uppercase;
    }

    ul {
      display: none;
      margin-top: 0;
      padding-top: $margin;
      padding-left: $margin;
      padding-bottom: $margin;

      &:hover {
        display: block;
      }
    }

    h2:hover + ul {
      display: block;
    }
  }

  .hamburger {
    display: none;
  }

  .socials {
    position: fixed;
    left: auto;
    bottom: $margin;
    right: $margin;
    padding-right: 0;
    margin-right: 0;
  }
}
