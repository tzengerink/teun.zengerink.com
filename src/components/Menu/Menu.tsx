import classNames from 'classnames'
import React from 'react'
import { Project } from '../../lib/projects'
import MenuItem from '../MenuItem/MenuItem'

interface MenuProps {
  projects: Project[]
  isOpen: boolean
  onOpen: () => void
  onItemClick: () => void
}

const Menu: React.FC<MenuProps> = ({ projects, isOpen, onOpen, onItemClick }) => {
  return (
    <nav
      data-testid="menu"
      className={classNames(
        'group',
        'fixed',
        'z-30',
        'inset-y-0',
        'left-xs',
        'w-full',
        'p-md',
        'pt-xl',
        'bg-transparant-white',
        'ease-linear',
        'duration-100',
        { 'translate-y-0': isOpen, 'opacity-100': isOpen, visible: isOpen },
        { 'translate-y-[-20px]': !isOpen, 'opacity-0': !isOpen, invisible: !isOpen },
        'md:translate-y-0',
        'md:translate-y-[-20px]',
        'md:visible',
        'md:opacity-100',
        'md:top-md',
        'md:right-md',
        'md:bottom-auto',
        'md:left-auto',
        'md:w-auto',
        'md:p-0',
        'md:pt-md',
        'md:text-right',
      )}
    >
      <h2 className={classNames('my-xs', 'uppercase', 'text-xs', 'text-grey/80', 'cursor-default')} onClick={onOpen}>
        Projects
      </h2>
      <ul
        className={classNames(
          'leading-5',
          { invisible: !isOpen },
          { 'ease-linear': !isOpen },
          { 'duration-100': !isOpen },
          { 'translate-y-[-20px]': !isOpen },
          { 'group-hover:visible': !isOpen },
          { 'group-hover:translate-y-0': !isOpen },
        )}
      >
        {projects.map((project) => (
          <MenuItem key={project.slug} project={project} onClick={onItemClick} />
        ))}
      </ul>
    </nav>
  )
}

export default Menu
