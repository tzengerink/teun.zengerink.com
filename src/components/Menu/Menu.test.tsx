import { fireEvent, render } from '@testing-library/react'
import React from 'react'
import mockProjects from '../../__mocks__/projects'
import Menu from './Menu'

const mockRouter = jest.fn(() => ({ asPath: '/' }))
jest.mock('next/router', () => ({ useRouter: () => mockRouter() }))

const defaultProps = { pageTitle: 'My Homepage', projects: mockProjects }

beforeEach(jest.clearAllMocks)

describe('Menu', () => {
  const renderComponent = (props = defaultProps) => render(<Menu {...props} />)

  it('renders correctly when on homepage', () => {
    const { container } = renderComponent()
    expect(container).toMatchSnapshot()
  })

  it('renders correctly when on a project page', () => {
    mockRouter.mockImplementation(() => ({ asPath: `${mockProjects[1].slug}` }))
    const { container } = renderComponent()
    expect(container).toMatchSnapshot()
  })

  it('filters out archived projects', () => {
    // TODO
  })

  it('toggles the menu when the hamburger is clicked', () => {
    const { getByTestId } = renderComponent()
    const menu = getByTestId('menu')
    const hamburger = getByTestId('hamburger')
    expect(menu.classList.contains('open')).toBe(false)
    fireEvent.click(hamburger)
    expect(menu.classList.contains('open')).toBe(true)
    fireEvent.click(hamburger)
    expect(menu.classList.contains('open')).toBe(false)
  })
})
