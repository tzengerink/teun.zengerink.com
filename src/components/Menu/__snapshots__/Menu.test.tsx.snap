// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu renders correctly 1`] = `
<div>
  <nav
    class="group fixed z-30 inset-y-0 left-xs w-full p-md pt-xl bg-transparant-white ease-linear duration-100 translate-y-[-20px] opacity-0 invisible md:translate-y-0 md:translate-y-[-20px] md:visible md:opacity-100 md:top-md md:right-md md:bottom-auto md:left-auto md:w-auto md:p-0 md:pt-md md:text-right"
    data-testid="menu"
  >
    <h2
      class="my-xs uppercase text-xs text-grey/80 cursor-default"
    >
      Projects
    </h2>
    <ul
      class="leading-5 invisible ease-linear duration-100 translate-y-[-20px] group-hover:visible group-hover:translate-y-0"
    >
      <a>
        My First Project
      </a>
      <a>
        Project without a statement
      </a>
      <a>
        Archived project
      </a>
    </ul>
  </nav>
</div>
`;

exports[`Menu renders correctly when open 1`] = `
<div>
  <nav
    class="group fixed z-30 inset-y-0 left-xs w-full p-md pt-xl bg-transparant-white ease-linear duration-100 translate-y-0 opacity-100 visible md:translate-y-0 md:translate-y-[-20px] md:visible md:opacity-100 md:top-md md:right-md md:bottom-auto md:left-auto md:w-auto md:p-0 md:pt-md md:text-right"
    data-testid="menu"
  >
    <h2
      class="my-xs uppercase text-xs text-grey/80 cursor-default"
    >
      Projects
    </h2>
    <ul
      class="leading-5"
    >
      <a>
        My First Project
      </a>
      <a>
        Project without a statement
      </a>
      <a>
        Archived project
      </a>
    </ul>
  </nav>
</div>
`;
